docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md

```diff
@@ -0,0 +1,274 @@
+whole subvolume first, and only then perform the binds to target paths.
+
+#### For case (a)
+
+Subvolume paths are normally retrieved by
+`ceph fs subvolume info/getpath <VOLUME NAME> <SUBVOLUME NAME> <SUBVOLUMEGROUP NAME>`,
+which outputs a path like so:
+
+```
+/volumes/<VOLUME NAME>/<SUBVOLUME NAME>/<UUID>
+```
+
+Snapshots are then accessible in:
+
+* `/volumes/<VOLUME NAME>/<SUBVOLUME NAME>/.snap` and
+* `/volumes/<VOLUME NAME>/<SUBVOLUME NAME>/<UUID>/.snap`.
+
+`/volumes/<VOLUME NAME>/<SUBVOLUME NAME>/<UUID>` may be deleted if the source
+subvolume is deleted, but thanks to the `snapshot-retention` feature, snapshots
+in `/volumes/<VOLUME NAME>/<SUBVOLUME NAME>/.snap` will remain to be available.
+
+The CephFS mount should therefore have its root set to the parent of what
+`fs subvolume getpath` returns, i.e. `/volumes/<VOLUME NAME>/<SUBVOLUME NAME>`.
+That way we will have snapshots available regardless of whether the subvolume
+itself still exists or not.
+
+#### For case (b)
+
+For cases where subvolumes are managed externally and not by cephfs-csi, we
+must assume that the cephx user we're given can access only
+`/volumes/<VOLUME NAME>/<SUBVOLUME NAME>/<UUID>` so users won't be able to
+benefit from snapshot retention. Users will need to be careful not to delete
+the parent subvolumes and snapshots while they are associated by these shallow
+RO volumes.
+
+### `NodePublishVolume`, `NodeUnpublishVolume`
+
+Node publish is trivial. We bind staging path to target path as a read-only
+mount.
+
+### `NodeGetVolumeStats`
+
+`NodeGetVolumeStatsResponse.usage[*].available` should be always zero.
+
+## Volume parameters, volume context
+
+This section provides a discussion around determinig what volume parameters and
+volume context parameters will be used to convey necessary information to the
+cephfs-csi driver in order to support shallow volumes.
+
+Volume parameters `CreateVolumeRequest.parameters`:
+
+* Should be "shallow" the default mode for all `CreateVolume` calls that have
+  (a) snapshot as data source and (b) read-only volume access mode? If not,
+  a new volume parameter should be introduced: e.g `isShallow: <bool>`. On the
+  other hand, does it even makes sense for users to want to create full copies
+  of snapshots and still have them read-only?
+
+Volume context `Volume.volume_context`:
+
+* Here we definitely need `isShallow` or similar. Without it we wouldn't be
+  able to distinguish between a regular volume that just happens to have
+  a read-only access mode, and a volume that references a snapshot.
+* Currently cephfs-csi recognizes `subvolumePath` for dynamically provisioned
+  volumes and `rootPath` for pre-previsioned volumes. As mentioned in
+  [`NodeStageVolume`, `NodeUnstageVolume` section](#NodeStageVolume-NodeUnstageVolume),
+  snapshots cannot be mounted directly. How do we pass in path to the parent
+  subvolume?
+  * a) Path to the snapshot is passed in via `subvolumePath` / `rootPath`,
```

Comment:
to keep things simple let's have an extra path for snapshot also and this will be used only in the NodePublish right?

Reply:
:+1: yes

Reply:
When do we look up the path of the snapshot itself?

The dirname has following format: `_<SNAP NAME>_<INODE NUMBER>`. We don't know what it is the directory called unless we actually iterate through `.snap` and find the correct path, i.e. the directory whose name contains `<SNAP NAME>`.

https://github.com/ceph/ceph/blob/a17ebc0406ac5592911fa7f596464c9c848a9c1e/src/mds/snap.cc#L81-L91

When?
* a) In `CreateVolume`. We'd need to mount the volume, look up the snapshot dir, and then unmount the volume. Dirname would be then stored in the response, in `Volume.volume_context["snapshotDirname"]`.
   * Pros: lookup is done only once per volume.
   * Cons: pre-provisioned volumes would have to do the lookup themselves in order for them to populate `Volume.volume_context["snapshotDirname"]`.
* b) In `NodeStageVolume`. `Volume.volume_context["snapshotName"]` must be populated. In this case we'd also need to mount the volume, but we'd need to do that either way. The only extra bit of work is doing the lookup once per PVC per node.
   * Pros: dynamically provisioned and pre-provisioned volumes are treated the same way.
   * Cons: one extra directory lookup once per PVC per node. Probably negligible performance penalty.
* c) Combination of a) and b). `CreateVolume` would populate `Volume.volume_context["snapshotDirname"]`. If it's not present in volume context during `NodeStageVolume` (in case of pre-provisioned volumes), but `snapshotName` is, it would do the lookup then.

Reply:
If the subvolume path is `csi-vol-1d17e149-f522-11eb-938d-866c88286c89/e50a2a89-a4fb-4fca-b590-8e8476e8a1a6/` and the snapshot path is present at `csi-vol-1d17e149-f522-11eb-938d-866c88286c89/.snap/csi-snap-f0a7e7be-f524-11eb-938d-866c88286c89` with this we can mount the volume at the stagingPath? what do you think?

Reply:
cc @kotreshhr

Reply:
Ah right, snapshot naming is different in parent dir. Thanks.
```
# pwd
.../volumes/csi/csi-vol-9dba52f1-f54b-11eb-b0aa-0242ac110006
# ls .snap
csi-snap-b7063583-f54b-11eb-b0aa-0242ac110006
```

The problem is when provisioning is managed by an external service (e.g. OpenStack Manila). In such environments the cephx clients have access restricted to the subvolume path only. `.snap` in parent is inaccessible, so they'd need to go to `subvol/.snap`:
```
# pwd
.../volumes/csi/csi-vol-9dba52f1-f54b-11eb-b0aa-0242ac110006/e6bb7726-71f3-4d32-940c-c34ea08ed383
# ls .snap
_csi-snap-b7063583-f54b-11eb-b0aa-0242ac110006_2199023257056
```
And here you can see the `_<SNAP NAME>_<INODE NUMBER>` name format I mentioned in my comment above. Since we don't know the ino num beforehand, *somebody* needs to do a name lookup.

Reply:
So for volumes provisioned by cephfs-csi, dirname is trivial to figure out.

Who should do snap dirname lookup for pre-provisioned volumes / volumes managed externally:
* a) The consumer. cephfs-csi would take dirname via volume context. This means that the consumer needs to mount the volume, perform the lookup, unmount it, and then create a pre-provisioned shallow volume with dirname specified.
* b) cephfs-csi. Since cephfs-csi already needs to mount the whole volume, it could easily do the dirname lookup itself. In this case users would only need to supply the snapshot name.

Which do you think makes more sense? b) is obviously a lot more user-friendly.

Reply:
for pre-provisioned volume, i think we can go for `option a` to keep it simple and as the user is manually creating it he can populate the volumeContext with required snap details so cephcsi will mount it